packages/iov-encoding/src/decimal.ts

```diff
@@ -0,0 +1,86 @@
+import BN from "bn.js";
+
+/**
+ * A type for arbitrary precision, non-negative decimals.
+ *
+ * Instances of this class are immutable.
+ */
+export class Decimal {
+  public static fromUserInput(input: string, fractionalDigits: number): Decimal {
+    if (!Number.isInteger(fractionalDigits)) throw new Error("Fractional digits is not an integer");
+    if (fractionalDigits < 0) throw new Error("Fractional digits must not be negative");
+
+    for (let index = 0; index < input.length; index++) {
+      if (!input[index].match(/^[0-9.]$/)) {
+        throw new Error(`Invalid character at position ${index + 1}`);
+      }
+    }
+
+    let whole: string;
+    let fractional: string;
+
+    if (input.search(/\./) === -1) {
+      // integer format, no separator
+      whole = input;
+      fractional = "";
+    } else {
+      const parts = input.split(".");
```

Comment:
You could avoid the `if`/`else` with something like this:
```ts
const parts = input.split(".");
if (parts.length === 0 || parts.length > 2) throw new Error("...");
const [whole, rawFractional = ""] = parts;
const fractional = rawFractional.replace(/0+$/, "");
```

Reply:
I used this code befor. However, this made "." a valid input, which I wanted to avoid. Then I decided to write it a little more explicit.